provider/datagen/src/transform/cldr/datetime/neo.rs

```diff
@@ -383,11 +462,156 @@ fn months_convert(
+        let pattern = pattern
+            .as_ref()
+            .expect("Failed to apply a coarse hour cycle to a full pattern.");
+        pattern.into()
+    } else {
+        (&pattern).into()
+    };
+    Ok(TimePatternV1 { pattern })
+}
+
+/// Looks at the hour cycle in `time_pattern`, and return the subtag related to the *opposite* hour cycle; i.e. the
+/// non-default one
+fn nondefault_subtag(
+    time_pattern: &ca::LengthPattern,
+    subtag12: Subtag,
+    subtag24: Subtag,
+) -> Subtag {
+    let pattern = time_pattern.get_pattern();
+
+    let pattern = pattern
+        .parse()
+        .expect("Failed to create a Pattern from bytes.");
+
+    let hc = CoarseHourCycle::determine(&pattern)
+        .expect("Could not find preferred hour cycle in locale");
+    match hc {
+        // this must invert the hour cycle since we're getting the nondefault one
+        CoarseHourCycle::H11H12 => subtag24,
+        CoarseHourCycle::H23H24 => subtag12,
+    }
+}
+
+// Time patterns have a manual implementation since they have custom supported_locales logic below
+impl DataProvider<TimePatternV1Marker> for DatagenProvider {
+    fn load(&self, req: DataRequest) -> Result<DataResponse<TimePatternV1Marker>, DataError> {
+        self.load_neo_patterns_key::<TimePatternV1Marker>(
+            req,
+            value!("gregory"),
+            timepattern_convert,
+        )
+    }
+}
+
+// Potential future optimization: if we split out aux subtags from supported_locales into a separate
+// method that eagerly generates all aux subtags (even if unused) and expects load() to figure out if the aux
+// subtag actually should be produced (by returning a special error), then this code is no longer necessary
+// and we can use a union of the H12/H24 key lengths arrays, instead checking for preferred hc
+// in timepattern_convert
+impl IterableDataProvider<TimePatternV1Marker> for DatagenProvider {
+    fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
+        let calendar = value!("gregory");
+        let mut r = Vec::new();
+
+        let cldr_cal = supported_cals()
+            .get(&calendar)
+            .ok_or_else(|| DataErrorKind::MissingLocale.into_error())?;
+        r.extend(self.cldr()?.dates(cldr_cal).list_langs()?.flat_map(|lid| {
+            let data = self
+                .load_calendar_dates(&lid, &calendar)
+                .expect("list_langs returned a language that couldn't be loaded");
+            let tp = &data.time_formats;
+
+            let keylengths = [
+                PATTERN_FULL,
+                PATTERN_LONG,
+                PATTERN_MEDIUM,
+                PATTERN_SHORT,
+                nondefault_subtag(&tp.full, PATTERN_FULL12, PATTERN_FULL24),
+                nondefault_subtag(&tp.long, PATTERN_LONG12, PATTERN_LONG24),
+                nondefault_subtag(&tp.medium, PATTERN_MEDIUM12, PATTERN_MEDIUM24),
+                nondefault_subtag(&tp.short, PATTERN_SHORT12, PATTERN_SHORT24),
```

Comment:
Thought: We want `en-GB-x-f24` to fallback to `en-GB-x-f` before `en-x-f24`. Same issue about horizontal fallback that we've encountered elsewhere. I have some thoughts on how to solve this. We can do it after landing this PR.